src/swarm-lang/Swarm/Language/Elaborate.hs

```diff
@@ -27,31 +31,72 @@ elaborate =
 
-  rewriteTerm :: Term' Polytype -> Term' Polytype
+  rewriteTerm :: TTerm -> TTerm
   rewriteTerm = \case
     -- For recursive let bindings, rewrite any occurrences of x to
     -- (force x).  When interpreting t1, we will put a binding (x |->
     -- delay t1) in the context.
-    SLet True x ty t1 t2 -> SLet True x ty (wrapForce (lvVar x) t1) (wrapForce (lvVar x) t2)
-    -- Rewrite any recursive occurrences of x inside t1 to (force x).
-    -- When a TDef is encountered at runtime its body will immediately
-    -- be wrapped in a VDelay. However, to make this work we also need
-    -- to wrap all free variables in any term with 'force' --- since
-    -- any such variables must in fact refer to things previously
-    -- bound by 'def'.
-    SDef True x ty t1 -> SDef True x ty (wrapForce (lvVar x) t1)
+    --
+    -- Here we also take inferred types for variables bound by let or
+    -- bind and stuff them into the term itself, so that we will still
+    -- have access to them at runtime, after type annotations on the
+    -- AST are erased.  We need them at runtime so we can keep track
+    -- of the types of variables in scope, for use in typechecking
+    -- additional terms entered at the REPL.
+    --
+    -- We assume requirements for these variables have already been
+    -- filled in during typechecking.  The reason we need to wait
+    -- until now to fill in the types is that we have to wait until
+    -- solving, substitution, and generalization are complete.
+    --
+    -- Eventually, once requirements analysis is part of typechecking,
+    -- we'll infer them both at typechecking time then fill them in
+    -- during elaboration here.
+    SLet ls r x mty mreq t1 t2 ->
+      let wrap
+            | r = wrapForce (lvVar x) -- wrap in 'force' if recursive
+            | otherwise = id
+       in SLet ls r x (mty <|> Just (t1 ^. sType)) mreq (wrap t1) (wrap t2)
+    SBind x (Just ty) _ mreq c1 c2 -> SBind x Nothing (Just ty) mreq c1 c2
     -- Rewrite @f $ x@ to @f x@.
     SApp (Syntax' _ (SApp (Syntax' _ (TConst AppF) _ _) l) _ _) r -> SApp l r
     -- Leave any other subterms alone.
     t -> t
 
-wrapForce :: Var -> Syntax' Polytype -> Syntax' Polytype
+wrapForce :: Var -> TSyntax -> TSyntax
 wrapForce x = mapFreeS x (\s@(Syntax' l _ ty cs) -> Syntax' l (SApp sForce s) ty cs)
 
 -- Note, TyUnit is not the right type, but I don't want to bother
 
-sForce :: Syntax' Polytype
+sForce :: TSyntax
 sForce = Syntax' NoLoc (TConst Force) Empty (Forall ["a"] (TyDelay (TyVar "a") :->: TyVar "a"))
+
+-- | Insert a special 'suspend' primitive at the very end of an erased
+--   term which must have a command type.
+insertSuspend :: Term -> Term
+insertSuspend t = case t of
+  -- Primitive things which have type Cmd Unit: p => (p ; suspend ())
+  TRequireDevice {} -> thenSuspend
+  TRequire {} -> thenSuspend
+  TRequirements {} -> thenSuspend
+  -- Recurse through let, tydef, bind, and annotate (but NOT through
+  -- let).
```

Comment:
This "recurse through let but not let" seems to be missing something, perhaps:
```suggestion
  -- the let body).
```

Reply:
Thanks for spotting that.  The incorrect comment made me realize that this code was actually wrong.   I was trying to make it so that `let` makes truly local bindings which are only available in its lexical scope, whereas `def` makes "global" bindings which are available in later terms entered at the REPL.  However, not recursing through `let` here was the wrong way to accomplish this.  I've now fixed it in a better way.

```
> let y = 3 in y + 1
it0 : Int = 4
> y
1:1: Unbound variable y
> def x = 3 end; return (x + 1)
it1 : Int = 4
> x
it2 : Int = 3
```